File: task310/src/main/java/com/example/task310/service/WriterService.java
package com.example.task310.service;

import com.example.task310.domain.Writer;
import com.example.task310.dto.request.WriterRequestTo;
import com.example.task310.dto.response.WriterResponseTo;
import com.example.task310.error.NotFoundException;
import com.example.task310.repo.WriterRepo;
import org.springframework.stereotype.Service;

@Service
public class WriterService {

    private final WriterRepo repo;

    public WriterService(WriterRepo repo) {
        this.repo = repo;
    }

    public WriterResponseTo create(WriterRequestTo r) {
        var created = repo.create(new Writer(null, r.login(), r.password(), r.firstname(), r.lastname()));
        return toDto(created);
    }

    public java.util.List<WriterResponseTo> getAll() {
        return repo.findAll().stream().map(this::toDto).toList();
    }

    public WriterResponseTo getById(long id) {
        var w = repo.find(id).orElseThrow(() -> new NotFoundException("Writer not found: " + id));
        return toDto(w);
    }

    public WriterResponseTo update(long id, WriterRequestTo r) {
        if (!repo.exists(id)) throw new NotFoundException("Writer not found: " + id);
        var updated = repo.update(id, new Writer(id, r.login(), r.password(), r.firstname(), r.lastname()));
        return toDto(updated);
    }

    public void delete(long id) {
        if (!repo.exists(id)) throw new NotFoundException("Writer not found: " + id);
        repo.delete(id);
    }

    private WriterResponseTo toDto(Writer w) {
        return new WriterResponseTo(w.id(), w.login(), w.firstname(), w.lastname());
    }
}

File: 351002/Yakhnovets/rest-task/src/writer/service/writer.service.ts
import { ForbiddenException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { WriterRequestTo } from '../dto/writer-request.to';
import { WriterResponseTo } from '../dto/writer-response.to';
import { Writer } from '../entity/writer.entity';
import { QueryFailedError, Repository } from 'typeorm';
import { ReactionService } from '../../reaction/service/reaction.service';

@Injectable()
export class WriterService {
  constructor(
    @InjectRepository(Writer)
    private readonly repository: Repository<Writer>,
    private readonly reactionService: ReactionService,
  ) {}

  async getAll(): Promise<WriterResponseTo[]> {
    const writers = await this.repository.find();
    return writers.map((writer) => this.toResponse(writer));
  }

  async getById(id: number): Promise<WriterResponseTo> {
    const writer = await this.repository.findOne({ where: { id } });
    if (!writer) throw new NotFoundException(`Writer with id=${id} not found`);
    return this.toResponse(writer);
  }

  async create(dto: WriterRequestTo): Promise<WriterResponseTo> {
    const entity = this.repository.create(this.toEntity(dto));
    try {
      const created = await this.repository.save(entity);
      return this.toResponse(created);
    } catch (error) {
      this.handleDbError(error);
    }
  }

  async update(id: number, dto: WriterRequestTo): Promise<WriterResponseTo> {
    const existing = await this.repository.findOne({ where: { id } });
    if (!existing) throw new NotFoundException(`Writer with id=${id} not found`);

    const updatedEntity = this.repository.merge(existing, this.toEntity(dto));
    try {
      const updated = await this.repository.save(updatedEntity);
      return this.toResponse(updated);
    } catch (error) {
      this.handleDbError(error);
    }
  }

  async delete(id: number): Promise<void> {
    const writer = await this.repository.findOne({
      where: { id },
      relations: { issues: true },
    });
    if (!writer) throw new NotFoundException(`Writer with id=${id} not found`);
    for (const issue of writer.issues ?? []) {
      await this.reactionService.deleteByIssueId(issue.id);
    }
    await this.repository.remove(writer);
  }

  private handleDbError(error: unknown): never {
    if (error instanceof QueryFailedError) {
      const driverError = error.driverError as { code?: string };
      if (driverError?.code === '23505') {
        throw new ForbiddenException('Writer login must be unique');
      }
    }

    throw error;
  }

  private toEntity(dto: WriterRequestTo): Omit<Writer, 'id' | 'issues'> {
    return {
      login: dto.login,
      password: dto.password,
      firstname: dto.firstname,
      lastname: dto.lastname,
    };
  }

  private toResponse(entity: Writer): WriterResponseTo {
    return {
      id: entity.id,
      login: entity.login,
      password: entity.password,
      firstname: entity.firstname,
      lastname: entity.lastname,
    };
  }
}
